src/Microsoft.ML.Data/Prediction/CalibratorCatalog.cs

```diff
@@ -0,0 +1,438 @@
+                        throw Host.Except($"{expected} column '{column.Name}' is not found");
+                    if (!column.IsCompatibleWith(outCol))
+                        throw Host.Except($"{expected} column '{column.Name}' is not compatible");
+                }
+            };
+
+            // check the input schema
+            checkColumnValid(ScoreColumn, DefaultColumnNames.Score);
+            checkColumnValid(WeightColumn, DefaultColumnNames.Weight);
+            checkColumnValid(LabelColumn, DefaultColumnNames.Label);
+            checkColumnValid(FeatureColumn, DefaultColumnNames.Features);
+            checkColumnValid(PredictedLabel, DefaultColumnNames.PredictedLabel);
+
+            //create the new Probability column
+            var outColumns = inputSchema.ToDictionary(x => x.Name);
+            outColumns[DefaultColumnNames.Probability] = new SchemaShape.Column(DefaultColumnNames.Probability,
+                SchemaShape.Column.VectorKind.Scalar,
+                NumberType.R4,
+                false,
+                new SchemaShape(MetadataUtils.GetTrainerOutputMetadata(true)));
+
+            return new SchemaShape(outColumns.Values);
+        }
+
+        public CalibratorTransformer<TICalibrator> Fit(IDataView input)
+        {
+            TICalibrator calibrator = null;
+
+            var roles = new List<KeyValuePair<RoleMappedSchema.ColumnRole, string>>();
+            roles.Add(RoleMappedSchema.CreatePair(MetadataUtils.Const.ScoreValueKind.Score, DefaultColumnNames.Score));
+            roles.Add(RoleMappedSchema.ColumnRole.Label.Bind(LabelColumn.Name));
+            roles.Add(RoleMappedSchema.ColumnRole.Feature.Bind(FeatureColumn.Name));
+            if (WeightColumn.IsValid)
+                roles.Add(RoleMappedSchema.ColumnRole.Weight.Bind(WeightColumn.Name));
+
+            var roleMappedData = new RoleMappedData(input, opt: false, roles.ToArray());
+
+            using (var ch = Host.Start("Creating calibrator."))
+            {
+                calibrator = (TICalibrator)CalibratorUtils.TrainCalibrator(Host, ch, CalibratorTrainer, Predictor, roleMappedData);
+            }
+
+            return Create(Host, calibrator);
+        }
+
+        protected abstract CalibratorTransformer<TICalibrator> Create(IHostEnvironment env, TICalibrator calibrator);
+    }
+
+    /// <summary>
+    /// CalibratorTransfomers, the artifact of calling Fit on a <see cref="CalibratorEstimatorBase{TCalibratorTrainer, TICalibrator}"/>.
+    /// If you pass a scored data, to the <see cref="CalibratorTransformer{TICalibrator}"/> Transform method, it will add the Probability column
+    /// to the dataset. The Probability column is the value of the score normalized to be a valid probability.
+    /// </summary>
+    /// <typeparam name="TICalibrator"></typeparam>
+    public abstract class CalibratorTransformer<TICalibrator> : RowToRowTransformerBase, ISingleFeaturePredictionTransformer<TICalibrator>
```

Comment:
If one day, hinge loss is added in FFM, it will not able to use this `CalibratorTransformer` because of `ISingleFeaturePredictionTransformer`. A calibrator just maps a scalar to another scalar, I am not sure why `ISingleFeaturePredictionTransformer` is needed.

Reply:
Could you please explain this concept in its doc string? Just saying that `CalibratorTransform` is a trainable model which maps the raw score to probability; that is, score can be viewed as feature in `ISingleFeaturePredictionTransformer` while probability is treated as the `label`.

---
In reply to: [240421876](https://github.com/dotnet/machinelearning/pull/1816#discussion_r240421876) [](ancestors = 240421876,240339828)